src/uct/ib/base/ib_device.c

```diff
@@ -482,6 +482,66 @@ void uct_ib_handle_async_event(uct_ib_device_t *dev, uct_ib_async_event_t *event
     ucs_log(level, "IB Async event on %s: %s", uct_ib_device_name(dev), event_info);
 }
 
+static ucs_status_t uct_ib_device_get_path_buffer(uct_ib_device_t *dev,
+                                                  char *path_buffer)
+{
+    char *resolved_path;
+
+    resolved_path = realpath(dev->ibv_context->device->ibdev_path, path_buffer);
+    if (resolved_path == NULL) {
+        return UCS_ERR_IO_ERROR;
+    }
+
+    /* Make sure there is "/infiniband/" substring in path_buffer */
+    if (strstr(path_buffer, "/infiniband/") == NULL) {
+        return UCS_ERR_IO_ERROR;
+    }
+
+    return UCS_OK;
+}
+
+static ucs_status_t uct_ib_device_get_sf_ids(uct_ib_device_t *dev)
+{
+    long vendor_id, device_id;
+    char path_buffer[PATH_MAX];
+    char pci_bus[UCT_IB_PCI_BUS_ID_LEN + 1];
+    char *bus_id;
+    ucs_status_t status;
+
+    /* realpath name is of form /sys/devices/.../0000:03:00.0/<UUID>/infiniband/mlx5_0
+     * and bus_id is constructed from 0000:03:00.0 */
+
+    status = uct_ib_device_get_path_buffer(dev, path_buffer);
+    if (status != UCS_OK) {
+        return status;
+    }
+
+    bus_id = basename(dirname(dirname(dirname(path_buffer))));
+
+    /* get 0000:03 from 0000:03:00.0*/
+    memcpy(pci_bus, bus_id, UCT_IB_PCI_BUS_ID_LEN);
+    pci_bus[UCT_IB_PCI_BUS_ID_LEN] = 0;
+
+    /* /sys/class/pci_bus/0000:03/device/0000:03:00.0/vendor */
+    status = ucs_read_file_number(&vendor_id, 1, UCT_IB_DEVICE_SYSFS_PCI_BUS_FMT,
+                                  pci_bus, bus_id, "vendor");
+    if (status != UCS_OK) {
+        return status;
+    }
+
+    /* /sys/class/pci_bus/0000:03/device/0000:03:00.0/device */
+    status = ucs_read_file_number(&device_id, 1, UCT_IB_DEVICE_SYSFS_PCI_BUS_FMT,
+                                  pci_bus, bus_id, "device");
+    if (status != UCS_OK) {
```

Comment:
can we unite common code with 549-552?

Reply:
@yosefe Tried, but seems not simply to do it. The SYSFS path format is much different for both places, the contents and path layers are different. Need much more code to deal with the diffrence.

Reply:
@yosefe Fixed.